luaui/Widgets/unit_factory_assist_fix.lua

```diff
@@ -0,0 +1,122 @@
+local widget = widget ---@type Widget
+
+function widget:GetInfo()
+	return {
+		name      = "Factory Assist Fix",
+		desc      = "Fixes factory assist so that builders don't leave to repair damaged finished units",
+		author    = "TheDujin",
+		date      = "Jun 30 2025",
+		license   = "GNU GPL, v2 or later",
+		layer     = 0,
+		enabled   = true
+	}
+end
+
+----------------------------------------------------------------
+-- Globals
+----------------------------------------------------------------
+local myTeam = Spring.GetMyTeamID()
+
+-- key: unit ID of an assist-capable builder that my team owns/owned.
+-- value: True if my team still owns this unit, and it's still alive. False otherwise
+local myAssistBuilders = {}
+
+local isAssistBuilder = {}
+
+----------------------------------------------------------------
+-- Speedups
+----------------------------------------------------------------
+local spGetUnitCurrentCommand = Spring.GetUnitCurrentCommand
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+local spAreTeamsAllied = Spring.AreTeamsAllied
+local spGetUnitHealth = Spring.GetUnitHealth
+local spGetUnitDefID = Spring.GetUnitDefID
+local spGetUnitTeam = Spring.GetUnitTeam
+
+local CMD_REPAIR = CMD.REPAIR
+local CMD_GUARD = CMD.GUARD
+local CMD_REMOVE = CMD.REMOVE
+local CMD_OPT_ALT = CMD.OPT_ALT
+
+for unitDefID, unitDef in pairs(UnitDefs) do
+	if unitDef.isBuilder and unitDef.canAssist then
+		isAssistBuilder[unitDefID] = true
+	end
+end
+
+-- If this builder unit is repairing the newly built unit when it should
+-- instead be guarding the factory, remove the repair command
+local function maybeRemoveRepairCmd(builderUnitID, builtUnitID, factID)
+	local firstCmdID, _, _, firstCmdParam_1 = spGetUnitCurrentCommand(builderUnitID, 1)
+	local secondCmdID, _, _, secondCmdParam_1 = spGetUnitCurrentCommand(builderUnitID, 2)
+	if (firstCmdID == CMD_REPAIR
+		and secondCmdID == CMD_GUARD) then
+			local isRepairingBuiltUnit = firstCmdParam_1 == builtUnitID
+			local isGuardingFactory = secondCmdParam_1 == factID
+			if (isRepairingBuiltUnit and isGuardingFactory) then
+				spGiveOrderToUnit(builderUnitID, CMD_REMOVE, {firstCmdID}, CMD_OPT_ALT)
+			end
+	end
```

Comment:
Could reorder it a bit to save a function call and make the code less of a pyramid, the table in GiveOrder is also avoidable for 1-arg commands:
 ```suggestion
	local firstCmdID, _, _, firstCmdParam_1 = spGetUnitCurrentCommand(builderUnitID, 1)
	if firstCmdID ~= CMD_REPAIR or firstCmdParam_1 ~= builtUnitID then
		return
	end

	local secondCmdID, _, _, secondCmdParam_1 = spGetUnitCurrentCommand(builderUnitID, 2)
	if secondCmdID ~= CMD_GUARD or secondCmdParam_1 ~= factID then
		return
	end

	spGiveOrderToUnit(builderUnitID, CMD_REMOVE, firstCmdID, CMD_OPT_ALT)
```